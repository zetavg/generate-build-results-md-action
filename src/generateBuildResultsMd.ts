import { IOSBuildResult } from './schema'

export default function generateBuildResultsMd(
  buildResultsArray: unknown[]
): [string, string[]] {
  const errors: string[] = []
  const items: string[] = []

  buildResultsArray.forEach((buildResult: unknown, index: number) => {
    if (typeof buildResult !== 'object' || buildResult === null) {
      errors.push(
        `Invalid build result at index ${index}, expected Object, got ${buildResult}`
      )
      return
    }

    const buildResultObj = buildResult as Record<string, unknown>

    switch (buildResultObj.type) {
      case 'ios': {
        try {
          const iosBuildResult = IOSBuildResult.parse(buildResultObj)

          const versionAndBuildNumber = `${iosBuildResult.version} (${iosBuildResult.build_number})`

          if (
            iosBuildResult.archive_artifact_url &&
            iosBuildResult.testflight_upload_succeeded
          ) {
            items.push(
              `* **${iosBuildResult.name}**: \`${versionAndBuildNumber}\` has been uploaded to TestFlight. You can also download the Xcode archive [here](${iosBuildResult.archive_artifact_url}).`
            )
          } else if (iosBuildResult.testflight_upload_succeeded) {
            items.push(
              `* **${iosBuildResult.name}**: \`${versionAndBuildNumber}\` has been uploaded to TestFlight.`
            )
          } else if (iosBuildResult.archive_artifact_url) {
            items.push(
              `* **${iosBuildResult.name}**: Download the Xcode archive of \`${versionAndBuildNumber}\` [here](${iosBuildResult.archive_artifact_url}). This build has not been uploaded to TestFlight.`
            )
          }
        } catch (e) {
          errors.push(
            `Error processing iOS build result at index ${index}: ${e instanceof Error ? e.message : e}`
          )
          return
        }
        break
      }

      default: {
        errors.push(
          `Invalid build result type at index ${index}: ${buildResultObj.type}`
        )
        break
      }
    }
  })

  return [items.join('\n'), errors]
}
