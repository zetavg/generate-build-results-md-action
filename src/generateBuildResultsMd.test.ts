import generateBuildResultsMd from './generateBuildResultsMd'

describe('generateBuildResultsMd', () => {
  describe('error messages', () => {
    test('invalid schema', () => {
      const [, errorMessages] = generateBuildResultsMd([
        123,
        null,
        {},
        { type: 'ios' }
      ])

      expect(errorMessages[0]).toEqual(
        'Invalid build result at index 0, expected Object, got 123'
      )

      expect(errorMessages[1]).toEqual(
        'Invalid build result at index 1, expected Object, got null'
      )

      expect(errorMessages[2]).toEqual(
        'Invalid build result type at index 2: undefined'
      )

      expect(errorMessages[3]).toContain(
        'Error processing iOS build result at index 3'
      )
    })
  })

  describe('iOS build results', () => {
    test('do not have archive_artifact_url and not testflight_upload_succeeded', () => {
      const [buildResultsMd, errorMessages] = generateBuildResultsMd([
        {
          type: 'ios',
          name: 'iOS Sample',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: '',
          testflight_upload_succeeded: false
        }
      ])

      expect(errorMessages).toEqual([])
      expect(buildResultsMd).toBe('')
    })

    test('has archive_artifact_url and testflight_upload_succeeded', () => {
      const [buildResultsMd, errorMessages] = generateBuildResultsMd([
        {
          type: 'ios',
          name: 'iOS Sample',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: 'https://example.com',
          testflight_upload_succeeded: true
        }
      ])

      expect(errorMessages).toEqual([])
      expect(buildResultsMd).toMatchSnapshot()
    })

    test('has archive_artifact_url but not testflight_upload_succeeded', () => {
      const [buildResultsMd, errorMessages] = generateBuildResultsMd([
        {
          type: 'ios',
          name: 'iOS Sample',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: 'https://example.com',
          testflight_upload_succeeded: false
        }
      ])

      expect(errorMessages).toEqual([])
      expect(buildResultsMd).toMatchSnapshot()
    })

    test('testflight_upload_succeeded but does not have archive_artifact_url', () => {
      const [buildResultsMd, errorMessages] = generateBuildResultsMd([
        {
          type: 'ios',
          name: 'iOS Sample',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: 'https://example.com',
          testflight_upload_succeeded: false
        }
      ])

      expect(errorMessages).toEqual([])
      expect(buildResultsMd).toMatchSnapshot()
    })
  })

  describe('examples', () => {
    test('example 1', () => {
      const [buildResultsMd, errorMessages] = generateBuildResultsMd([
        {},
        {
          type: 'ios',
          name: 'iOS Sample - Archive',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: 'https://example.com',
          testflight_upload_succeeded: false
        },
        { type: 'ios' },
        {
          type: 'ios',
          name: 'iOS Sample - TestFlight',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: '',
          testflight_upload_succeeded: true
        },
        {
          type: 'ios',
          name: 'iOS Sample - Nothing',
          version: '1.2.3',
          build_number: 456,
          archive_artifact_url: '',
          testflight_upload_succeeded: false
        }
      ])

      expect(errorMessages).toMatchSnapshot()
      expect(buildResultsMd).toMatchSnapshot()
    })
  })
})
