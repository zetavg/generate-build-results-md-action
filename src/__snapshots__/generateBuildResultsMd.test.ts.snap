// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateBuildResultsMd examples example 1 1`] = `
[
  "Invalid build result type at index 0: undefined",
  "Error processing iOS build result at index 2: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "name"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "version"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "build_number"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "archive_artifact_url"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "undefined",
    "path": [
      "testflight_upload_succeeded"
    ],
    "message": "Required"
  }
]",
  "Error processing Android build result at index 6: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "name"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "version"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "apk_artifact_url"
    ],
    "message": "Required"
  }
]",
]
`;

exports[`generateBuildResultsMd examples example 1 2`] = `
"* **iOS Sample - Archive**: Download the Xcode archive of \`1.2.3 (456)\` [here](https://example.com). This build has not been uploaded to TestFlight.
* **iOS Sample - TestFlight**: \`1.2.3 (456)\` has been uploaded to TestFlight.
* **Android Sample - APK**: Download the APK file [here](https://example.com)."
`;

exports[`generateBuildResultsMd iOS build results has archive_artifact_url and testflight_upload_succeeded 1`] = `"* **iOS Sample**: \`1.2.3 (456)\` has been uploaded to TestFlight. You can also download the Xcode archive [here](https://example.com)."`;

exports[`generateBuildResultsMd iOS build results has archive_artifact_url but not testflight_upload_succeeded 1`] = `"* **iOS Sample**: Download the Xcode archive of \`1.2.3 (456)\` [here](https://example.com). This build has not been uploaded to TestFlight."`;

exports[`generateBuildResultsMd iOS build results testflight_upload_succeeded but does not have archive_artifact_url 1`] = `"* **iOS Sample**: Download the Xcode archive of \`1.2.3 (456)\` [here](https://example.com). This build has not been uploaded to TestFlight."`;
